File: src/Chapter04GraphBFS/Path.java
package Chapter04GraphBFS;

import Chapter04GraphBFS.Graph;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;

public class Path {
    private Graph G;
    private int s, t;

    private int[] pre;
    private boolean[] visited;

    public Path(Graph g, int s, int t) {


        g.validateVertex(s);
        g.validateVertex(t);
        this.G = g;
        this.s = s;
        this.t = t;
        visited = new boolean[G.V()];
        pre = new int[G.V()];
        for (int i = 0; i < pre.length; i++) {
            pre[i] = -1;
        }

        bfs();
        for (boolean e : visited)
            System.out.print(e + " ");
        System.out.println();
    }

    public void bfs() {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(s);
        visited[s] = true;
        pre[s] = s;
        if (s == t)
            return;
        while (!queue.isEmpty()) {
            int v = queue.remove();
            for (int w : G.adj(v)) {
                if (!visited[w]) {
                    visited[w] = true;
                    queue.add(w);
                    pre[w] = v;
                    if (w == t)
                        return;
                }
            }
        }
    }

    public boolean isConnected() {
        return visited[t];
    }

    public Iterable<Integer> path() {

        ArrayList<Integer> res = new ArrayList<>();
        if (!isConnected()) return res;

        int cur = t;
        while (cur != s) {
            res.add(cur);
            cur = pre[cur];
        }
        res.add(s);

        Collections.reverse(res);
        return res;
    }

    public static void main(String[] args) {

        Chapter04GraphBFS.Graph g = new Chapter04GraphBFS.Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g.txt");
        Chapter04GraphBFS.Path path = new Chapter04GraphBFS.Path(g, 0, 6);
        System.out.println("0 -> 6 : " + path.path());

        Chapter04GraphBFS.Path path2 = new Chapter04GraphBFS.Path(g, 0, 5);
        System.out.println("0 -> 5 : " + path2.path());

        Chapter04GraphBFS.Path path3 = new Chapter04GraphBFS.Path(g, 0, 1);
        System.out.println("0 -> 1 : " + path3.path());
    }
}

File: src/Chapter04GraphBFS/python/Path.py
from Chapter02GraphDFS.adjSet import adjSet as Graph
from collections import deque


class SingleSourcePath:
    def __init__(self, G, s, t):
        self._G = G
        self.t = t
        self._visited = [False] * G.V
        self._pre = [-1] * G.V
        self.s = s

        self._bfs(s)

    def _bfs(self, s):
        queue = deque()
        queue.append(s)
        self._pre[s] = s
        self._visited[s] = True

        while queue:
            v = queue.popleft()
            for w in self._G.adj(v):
                if not self._visited[w]:
                    queue.append(w)
                    self._visited[w] = True
                    self._pre[w] = v
                    if w == self.t:
                        return

    def path(self):
        if not self.is_connected_to() :
            return []
        w = self.t
        res = [w]
        while self._pre[w] != w:
            res.append(self._pre[w])
            w = self._pre[w]

        return list(reversed(res))

    def is_connected_to(self):
        self._G._validate_vertex(self.t)
        # if t is visited for current self._s
        # that implies t is in the path of self._s
        return self._visited[self.t]


if __name__ == "__main__":
    filename = "/Chapter03GraphDFSApplications/g.txt"
    g = Graph(filename)

    sspath = SingleSourcePath(g, 0, 6)
    print(sspath.path())
    print(sspath.is_connected_to())
